File: 백준/Gold/2143. 두 배열의 합/두 배열의 합.py
import sys
input = sys.stdin.readline
from collections import Counter

m = int(input())
c = int(input())
arr1 = list(map(int, input().split()))
d = int(input())
arr2 = list(map(int, input().split()))
answer = 0

## 구간별 누적함
def partSum(arr,n):
    sum_ = []
    n = len(arr)
    for i in range(0,n):
        for j in range(i+1, n+1):
            sum_.append(sum(arr[i:j]))
    return sum_




        
    
sum1=Counter(partSum(arr1, c))
sum2=Counter(partSum(arr2, d))
## print(sum1, sum2)

for i in sum1.keys():
    j = m-i
    if sum2[j]:
        answer+=(sum1[i]*sum2[j])


print(answer)

File: 백준/Gold/2143. 두 배열의 합/두 배열의 합.js
const readline = require("readline")
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
})

const lines = []


rl.on("line", (line) => {
    lines.push(line.trim())
}).on("close", () => {
    const N = Number(lines[0])
    const K = Number(lines[1])
    const arr1 = lines[2].split(" ").map(Number)
    const P = Number(lines[3])
    const arr2 = lines[4].split(" ").map(Number)
    const arr1Dic = new Map()
    let answer = 0
    // 구간 합 arr1
    for (let i = 0; i < K; i++) {
        let pre_sum = 0
        for (let j = i; j < K; j++) {
            pre_sum += arr1[j]
            const v = (arr1Dic.get(pre_sum) ?? 0) + 1
            arr1Dic.set(pre_sum, v)
        }
    }


    // 구간 합 arr2
        for (let i = 0; i < P; i++) {
        let pre_sum = 0
        for (let j = i; j < P; j++) {
            pre_sum += arr2[j]
            const v = (arr1Dic.get(N-pre_sum) ?? 0) 
            answer+=v
           
        }
    }
    //console.log("close", N, K, P)
    //console.log("close", arr1, arr2)
    console.log(answer)
})